### DISCUSSION TITLE
`reload` automatically if the last client disconnects from an sbt server

### DISCUSSION THREAD

**Discussion by robstoll** [sbt/sbt#6828] [Ideas]
See #6826 as background, I assumed that when I have disconnected all clients and reconnect, that a `reload` was done automatically because in non `-client` mode you see warnings like:
```
[warn] build source files have changed
[warn] modified files: 
[warn]   /home/robert/projects/test-scala3/build.sbt
[warn] Apply these changes by running `reload`.
[warn] Automatically reload the build when source changes are detected by setting `Global / onChangedBuildSource := ReloadOnSourceChanges`.
```

Thus I suggest that we reload automatically or show again the warnings. Thoughts?

**Comment by adpi2** [comment]
Thanks for starting the discussion, it is an interesting idea.

For the warnings, I think we should indeed show them to all the clients, the old ones and the new ones.

When it comes to automatic reload we have `Global / onChangedBuildSource := ReloadOnSourceChanges` but this does not really work with sbt server and client because it freezes the server if there is a compilation error in the build. We should fix that by restoring the previous state if the reload fails.

Should we also reload automatically when there is no more client. Yes I think it is a good idea and we should try experimenting it. It could be something like: when a first client connects, we submit a command that check if the build sources has changed and reload if necessary.

I would be happy to help on this topic. Would you like to contribute?

**Reply by robstoll** [reply]
I never contributed to sbt so I surely would need some help to get going but I can try to be a helping hand.

> it could be something like: when a first client connects, we submit a command that check if the build sources has changed and reload if necessary.

Personally, I would perform this upon last disconnect because this way the next connect will be again blazing fast :)

**Reply by adpi2** [reply]
> I never contributed to sbt so I surely would need some help to get going but I can try to be a helping hand.

That's good to know thanks.

> Personally, I would perform this upon last disconnect because this way the next connect will be again blazing fast :)

Sure but the build file can also change between the last disconnect and the first connection. Maybe it would be possible to constantly watch for build source changes tough.

**Reply by robstoll** [reply]
True and thinking about it again I guess it will most of the time be like this:
1. disconnect all clients 
2. change something in your build.sbt 
3. eventually connect again

I guess it would be good enough to start with `do it on first connect` and we could still improve from there.
